bigflow/build/dataflow/depscheck.py

```diff
@@ -0,0 +1,116 @@
+    matcher = re.compile(r"(.*?)(?:\[.*?\])?==(.*)")
+    return dict(
+        matcher.match(line.lower()).groups()
+        for line in requirements
+    )
+
+
+def load_beam_worker_preinstalled_dependencies(beam_version, py_version):
+    fname = f"beam{beam_version}_py{py_version}.txt"
+    txt = importlib.resources.read_text(bigflow.build.dataflow.workerdeps, fname)
+    return dict(
+        line.lower().split("==", 2)
+        for line in txt.splitlines()
+        if line.strip() and not line.startswith("#")
+    )
+
+
+def detect_dataflow_conflicts(req_path: Path):
+
+    req_path = req_path.with_suffix(".txt")
+    existing_pinfile = req_path.parent / "dataflow_pins.in"
+
+    requirements = bigflow.build.pip.read_requirements(req_path)
+    beam_version = detect_beam_version(requirements)
+    if not beam_version:
+        logger.debug("Beam is not used - don't perform worker dependencies conflict check")
+        return
+
+    py_version = detect_py_version()
+    existing_pins = existing_pinfile.read_text() if existing_pinfile.exists() else ""
+
+    logger.debug("Beam version is %s", beam_version)
+    logger.debug("Python version is %s", py_version)
+
+    try:
+        workerdeps = load_beam_worker_preinstalled_dependencies(beam_version, py_version)
+    except FileNotFoundError:
+        logger.error("Unsupported beam/python version: %s/%s", beam_version, py_version)
+        return
+
+    reqs = build_requirements_dict(requirements)
+    common_deps = set(reqs) & set(workerdeps)
+    conflicts = {
+        k: (reqs[k], workerdeps[k])
+        for k in common_deps
+        if reqs[k] != workerdeps[k]
+        and not re.search(rf"\W{re.escape(k)}\W", existing_pins)  # pin is ignored
+    }
+
+    return conflicts
+
+
+def check_beam_worker_dependencies_conflict(req_path: Path):
+    conflicts = detect_dataflow_conflicts(req_path)
+    if not conflicts:
+        logger.debug("No conflicts found - woow!")
+    else:
+        logger.error(
+            textwrap.dedent(f"""
+                ========================================
+                The Apache Beam SDKs and Dataflow workers depend on common third-party components which then import additional dependencies.
+                Your project is using different verions of those libraries, which may result in conflicts, bugs or unexpected behaviour.
+                More information https://cloud.google.com/dataflow/docs/concepts/sdk-worker-dependencies
+                Conflicted libraries:
+                %s
+
+                It is recommended to resolve conflicts by running:
+                > bigflow codegen pin-dataflow-requirements
+                ========================================
+            """),
+            "\n".join(
+                f" - {dep}: used {our_version}, beam has {worker_version}"
+                for dep, (our_version, worker_version) in conflicts.items()
+            ),
+        )
+
+
+def generate_dataflow_pins_file(req_path=None):
```

Comment:
Im not sure if pins is the best name. Maybe overrides, dependency_fixes? Or something that describes the purpose.

Reply:
AFAIK "pin" is a commonly used term. 
I found many links/documents which use "pinning dependencies" and/or "pin" for such meaning.
like https://before-you-ship.18f.gov/infrastructure/pinning-dependencies/

Reply:
What I wanted to say is that "pins" describe the solution to the problem of fixing dependencies. So the question is – should that function tell what problem it solves or how it solves the problem (open question)?

Reply:
If you about function name - yes, it might be something like `sync_requirements_with_dataflow_workers`, `fix_requirements_with_dataflow_workers` etc etc. Do you have something in mind?

Reply:
`sync_requirements_with_dataflow_workers ` is fine